File: RyuHeySu/GFS/1743.py
# 음식물 피하기
from collections import deque

m, n, k = map(int, input().split())
data = [ ['.'] * n for _ in range(m) ]
for _ in range(k) :
    x, y = map(int, input().split())
    data[x - 1][y - 1] = '#'

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]

def DFS(x, y) :
    stak = deque()
    stak.append((x, y))

    data[x][y] = '.'
    count = 1

    while stak :
        a, b = stak.pop()

        for i in range(4) :
            nx = a + dx[i]
            ny = b + dy[i]

            if nx >= 0 and nx < m and ny >= 0 and ny < n :
                if data[nx][ny] == '#' :
                    count += 1
                    data[nx][ny] = '.'
                    stak.append((nx, ny))
    return count

result = 0
for i in range(m) :
    for j in range(n) :
        if data[i][j] == '#' :
            result = max(result, DFS(i, j))

print(result)

File: HeoJiye/beakjoon/DFS/1743.js
// https://www.acmicpc.net/problem/1743
// type err!

const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "../test.txt";
const input = fs.readFileSync(filePath).toString();

const line = input.split("\n");
const [N, M, K] = line.shift().split(" ").map(Number);
const data = line.map((e) => e.split(" ").map(Number));

const board = [];
for (let i = 0; i < N; i++) {
  board[i] = [];
  for (let j = 0; j < M; j++) {
    board[i][j] = true;
  }
}
for (const [i, j] of data) {
  board[i - 1][j - 1] = false;
}

// dfs로 순회하며 면적 세기
function dfs(board, i, j) {
  if (i < 0 || j < 0 || i >= N || j >= M || board[i][j]) return 0;

  board[i][j] = true;
  return (
    dfs(board, i - 1, j) +
    dfs(board, i, j - 1) +
    dfs(board, i + 1, j) +
    dfs(board, i, j + 1) +
    1
  );
}

let max_count = 0;

for (let i = 0; i < N; i++) {
  for (let j = 0; j < M; j++) {
    max_count = Math.max(max_count, dfs(board, i, j));
  }
}

console.log(max_count);
